### ISSUE TITLE
include statement differs from PHP version

### ISSUE THREAD

**Issue by xikky** [twigjs/twig.js#536]
Hi,

I am using twig for wordpress and I'm calling templates both from PHP and from JS ajax. I have an issue with the include statment in such a way that the path for a template when using PHP is always relative to the /veiws directory, but on the other hand,  when using Ajax, the path for the template in the include statement is relative to the directory of the template being included from. Hence if you organise your templates in sub-directories, seems like I have to make a check if template is requested from PHP or Ajax to provide the appropriate path.

Is there a cleaner way to do this? I'd rather not do an if else statement for different include paths same template, in my template files.

Thanks!

**Comment by ericmorand** [comment]
That's true, TwigPHP filesystem loader always load relatively to one of the paths that has been registered. In your case, the path 'views' has been registered and then the loader will always try to resolve relative paths from there.

Twig.js filesystem loader on the other hand resolve relative paths from the path of the template that do the include. Not sure if it is by design or if it is a bug but it's not consistent with TwigPHP.

Maybe there is an option somewhere in Twig.js that will solve your problem, but in case this doesn't happen, there is a very elegant way to solve this in TwigPHP: overriding the `include `node vie a custom Twig extension and a node visitor. The idea is to have the path of the template to load resolved by the `include` node.

I'm sorry I don't have an example in PHP but here is, in JavaScript, the custom `include `node I use with Twing to solve the problem:

```javascript
class CustomNodeInclude extends TwingNodeInclude {
    addGetTemplate(compiler) {
        compiler
            .raw('(() => {\n')
            .indent()
            .write('let path = require(\'path\');\n')
            .write('let currentPath = this.getSourceContext().path;\n')
            .write('let templatePath = path.resolve(path.dirname(currentPath), ')
            .subcompile(this.getNode('expr'))
            .raw(');\n')
            .write('return this.loadTemplate(')
            .raw('templatePath')
            .raw(', ')
            .repr(this.getTemplateName())
            .raw(', ')
            .repr(this.getTemplateLine())
            .raw(');\n')
            .outdent()
            .write('})()')
        ;
    }
};
```

As you can see, I resolve the template path at runtime. This shouldn't too difficult to adapt to PHP and to your needs. Then, implement a node visitor and a custom extension. Here are the classes I use for this, once again in JavaScript:

```javascript
class CustomNodeVisitor extends TwingBaseNodeVisitor {
    doEnterNode(node, env) {
        if (node instanceof TwingNodeInclude) {
            return new CustomNodeInclude(
                node.getNode('expr'),
                node.hasNode('variables') ? node.getNode('variables') : null,
                node.getAttribute('only'),
                node.getAttribute('ignore_missing'),
                node.getTemplateLine(),
                node.getNodeTag()
            );
        }

        return node;
    }

    doLeaveNode(node, env) {
        return node;
    }

    getPriority() {
        return 0;
    }
};

class CustomExtension extends TwingExtension {
    getNodeVisitors() {
        return [
            new CustomNodeVisitor()
        ];
    }
};
```

Hope that helps.

**Comment by willrowe** [comment]
Closing this due to lack of an example demonstrating the issue. If you are still encountering this problem, please provide an example using [this template](https://codepen.io/pen?template=LYQawea) and re-open this issue.
